### PR TITLE
Add a reboot_gateway() funtion for actual Plugwise devices.

### PR DESCRIPTION
**Pull request by bouwew**
Part of implementing https://github.com/plugwise/plugwise-beta/issues/604

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved handling and validation of POST request responses for more accurate error handling.

- **Chores**
  - Updated project version to 0.38.0 in configuration files.

### CODE REVIEW

**Comment by bouwew** [comment]
Test-coverage still required.

File: plugwise/helper.py [hunk 1]
```diff
@@ -123,6 +123,14 @@ async def _request(
                 resp = await self._websession.get(
                     url, headers=use_headers, auth=self._auth
                 )
+            if method == "post":
```
Comment (CoMPaTech) [inline plugwise/helper.py:126]: Maybe we should consider (not for this PR!) a rewrite to match/case for this construct

File: plugwise/smile.py [hunk 2]
```diff
@@ -131,15 +133,21 @@ async def async_update(self) -> PlugwiseData:
...
+            await self.full_update_device()
+            self.get_all_devices()
+            if "heater_id" in self.gw_data:
+                self._heater_id = self.gw_data["heater_id"]
+                if "cooling_enabled" in self.gw_devices[self._heater_id]["binary_sensors"]:
+                    self._cooling_enabled = self.gw_devices[self._heater_id]["binary_sensors"]["cooling_enabled"]
 
-        return PlugwiseData(self.gw_data, self.gw_devices)
+            return PlugwiseData(self.gw_data, self.gw_devices)
+        except:
```
Comment (CoMPaTech) [inline plugwise/smile.py:146]: Should we except `KeyError` as you spotted? As indicated by codefactor a bare except shouldn't be there

File: plugwise/helper.py (hunk 1 above)
Comment (bouwew) [reply plugwise/helper.py:126]: Sure that's a good idea :)

File: plugwise/smile.py (hunk 2 above)
Comment (bouwew) [reply plugwise/smile.py]: Solved.

**Comment by bouwew** [comment]
@coderabbitai review

File: plugwise/smile.py [hunk 3]
```diff
@@ -149,6 +152,10 @@
         """Delete the active Plugwise Notification."""
         await self._request(NOTIFICATIONS, method="delete")
 
+    async def reboot_gateway(self) -> None:
+        """Reboot the Gateway."""
+        await self._request(GATEWAY_REBOOT, method="post")
```
[the remark this answers is not in the document]
Comment (bouwew) [reply plugwise/smile.py]: Implemented.

File: plugwise/__init__.py [hunk 4]
```diff
@@ -367,3 +377,7 @@
     async def delete_notification(self) -> None:
         """Delete the active Plugwise Notification."""
         await self._smile_api.delete_notification()
+
+    async def reboot_gateway(self) -> None:
+        """Reboot the Plugwise Gateway."""
+        await self._smile_api.reboot_gateway()
```
[the remark this answers is not in the document]
Comment (bouwew) [reply plugwise/__init__.py]: Impelemented.

**Comment by bouwew** [comment]
@coderabbitai review

File: tests/test_adam.py [hunk 5]
```diff
@@ -78,11 +78,13 @@ async def test_connect_adam_zone_per_device(self):
...
             notification_deletion = False  # pragma: no cover
-        except pw_exceptions.ResponseError:
+        except pw_exceptions.PlugwiseError:
             notification_deletion = True
-
         assert notification_deletion
 
+        reboot = await self.tinker_reboot(smile, unhappy=True)
+        assert not reboot
+
```
[the remark this answers is not in the document]
Comment (bouwew) [reply tests/test_adam.py:87]: Moved positive reboot test under same test.

**Comment by CoMPaTech** [comment]
> Outside diff range and nitpick comments (1)
> > plugwise/helper.py (1)> Line range hint `145-145`: Define the `LOGGER` variable.
> > > ```diff
> > > + from logging import getLogger
> > > + LOGGER = getLogger(__name__)
> > > ```
> > > The `LOGGER` variable is used throughout the file but is not defined. This will cause a runtime error. Define `LOGGER` at the beginning of the file to fix this issue.
> > > Also applies to: 156-156, 179-179, 184-184, 190-190, 197-197, 648-648, 650-650
> 
> Review details

Good catch by the rabbit, as we only include it it will never occur but might be good to add it

**Review by CoMPaTech — APPROVED** [review]
Nice!

**Pull request merged**